### ISSUE TITLE
DataReceived event: message_finished

### ISSUE THREAD

**Issue by Kriechi** [python-hyper/wsproto#76]
While improving our docs, I noticed the `message_finished` member of the `DataReceived` event.
Maybe someone could help me understand this - or confirm my idea:

This basically means that we currently fire `TextReceived` and `BinaryReceived` events with incomplete messages? And the user is responsible to concatenate them to end up with the full / complete websocket message?

For some reason, I did not expect this - I assumed these events will ALWAYS give me the full message.
In hyper-h2 the RequestReceived and ResponseReceived contain the full headers, because the HTTP body is transmitted separately.

I feel like we should change wsproto to only emit complete messages via the `TextReceived` and `BinaryReceived` events. Any opinions about this?

**Comment by njsmith** [comment]
Yes, this is intentional: the websocket protocol inherently allows message fragmentation – so e.g. a PING can arrive in the *middle* of a message – and also we have to be able to return partial messages to avoid wsproto requiring unbounded memory. If the low-level API returns the message as a stream of fragments, then it's straightforward for a higher-level API to buffer and reassemble them if it wants; but if the low-level API does the buffering/reassembly, it's impossible to build a streaming consumer on top of that.
